src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -364,7 +363,14 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                                     target.is_example() ||
                                     profile.test;
 
-                is_correct_dep && is_actual_dep
+                // If the dependency is optional, then we're only activating it
+                // if the corresponding feature was activated
+                let activated = !d.is_optional() ||
+                                self.resolve.features(pkg.package_id()).map(|f| {
```

Comment:
How so? Currently it returns `Option<&HashSet>`, but you're saying it should return `HashSet`? Certainly possible, just wanted to avoid the extra work on the calling side for now.